titanic.py

```diff
@@ -86,19 +86,58 @@ def conditional_mutual_information(p, j, *conditional_indices):
     return (conditional_shannon_entropy(np.sum(p, axis=j), *marginal_conditional_indices)
             - conditional_shannon_entropy(p, j, *conditional_indices))
 
+
 def maximum_energy_delta(bqm):
     """Compute conservative bound on maximum change in energy when flipping a single variable"""
     return max(abs(bqm.get_linear(i))
                + sum(abs(bqm.get_quadratic(i,j))
                      for j in bqm.iter_neighbors(i))
                for i in bqm.iter_variables())
 
-def run_demo():
+
+def mutual_information_bqm(dataset, features):
+    """Build a BQM that maximizes MI between survival and a subset of features"""
+    variables = ((feature, -mutual_information(prob(dataset[['survived', feature]].values), 1))
+                 for feature in features)
+    interactions = ((f0, f1, -conditional_mutual_information(prob(dataset[['survived', f0, f1]].values), 1, 2))
+                    for f0, f1 in itertools.permutations(features, 2))
+    return dimod.BinaryQuadraticModel(variables, interactions, 0, dimod.BINARY)
+
+
+def add_combination_penalty(bqm, k, penalty):
+    """Create a new BQM with an additional penalty biased towards k-combinations"""
+    kbqm = dimod.generators.combinations(bqm.variables, k, strength=penalty)
+    kbqm.update(bqm)
+    return kbqm
+
+
+def mutual_information_feature_selection(dataset, features):
+    """Run the MIFS algorithm on a D-Wave solver"""
+    
+    # Set up a QPU sampler that embeds to a fully-connected graph of all the variables
+    sampler = DWaveCliqueSampler()
+
+    # For each number of features, k, penalize selection of fewer or more features
+    selected_features = np.zeros((len(features), len(features)))
+
+    bqm = mutual_information_bqm(dataset, features)
+
+    # This ensures that the soltion will satisfy the constraints.
+    penalty = maximum_energy_delta(bqm)
+
+    for k in range(1, len(features) + 1):
+        kbqm = add_combination_penalty(bqm, k, penalty)
+        sample = sampler.sample(kbqm,
+                                label='Example - MI Feature Selection',
+                                num_reads=10000).first.sample
+        for fi, f in enumerate(features):
+            selected_features[k-1, fi] = sample[f]
+    return selected_features
+
+
+def run_demo(dataset):
     # Read the feature-engineered data into a pandas dataframe
     # Data obtained from http://biostat.mc.vanderbilt.edu/DataSets
```

Comment:
Looks like these comments about reading the data are out of place after this refactor

Reply:
done